src/connector/src/parser/mod.rs

```diff
@@ -468,6 +468,18 @@ pub trait ByteStreamSourceParser: Send + Debug + Sized + 'static {
     }
 }
 
+#[try_stream(ok = Vec<SourceMessage>, error = anyhow::Error)]
+async fn ensure_largest_at_rate_limit(stream: BoxSourceStream, rate_limit: u32) {
+    #[for_await]
+    for batch in stream {
+        let mut batch = batch?;
+        while batch.len() > rate_limit as usize {
+            yield batch.drain(..rate_limit as usize).collect();
+        }
+        yield batch;
+    }
+}
+
```

Comment:
Ensures that the source stream's batch sizes are smaller than rate_limit.
That way non-transaction sources can be rate limited.

Reply:
There is overhead using drain to copy back tail elements.
```
        // Drain will ptr::read out the values to remove.
        // When finished, remaining tail of the vec is copied back to cover
        // the hole, and the vector length is restored to the new length.
```
As it is in the critical perf path, I suggest to move out parts instead of drain.
```
#[try_stream(ok = Vec<SourceMessage>, error = anyhow::Error)]
async fn ensure_largest_at_rate_limit(stream: BoxSourceStream, rate_limit: u32) {
    #[for_await]
    for batch in stream {
        let batch = batch?;
        let mut start = 0;
        let end = batch.len();
        while start < end {
            let next = std::cmp::min(start + rate_limit as usize, end);
            yield std::mem::take(batch[start..next].as_mut()).to_vec();
            start = next;
        }
    }
}
```

Reply:
I guess it may not be a major problem?

- In most of time when `rate_limit` is unset or set to a extremely large value, the `drain` branch won't be executed.

- If the `rate_limit` is small, there must be other more significant bottlenecks in the system. The overhead here pales in comparison.